### PR TITLE
Expands semantic_text tutorial with hybrid search

### PR DESCRIPTION
**Pull request by kosabogi**
## Overview

This PR expands the existing semantic_text tutorial by adding hybrid search method to it. 

- Changed the structure of the tutorial.
- Added detailed instructions for hybrid search, including index mapping creation, reindexing, and performing hybrid search queries.

## Related Issue
https://github.com/elastic/search-docs-team/issues/191

## Preview
Tutorial: semantic search and hybrid search with semantic_text

### CODE REVIEW

File: docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc [hunk 1]
```diff
@@ -172,10 +252,56 @@ POST _tasks/<task_id>/_cancel
...
+<1> The source index from which data will be reindexed.
+<2> The batch size is reduced to 10 for quicker processing and progress tracking. The default is 1000.
+<3> The destination index where the reindexed data will be ingested, in this case, the `semantic-hybrid-embeddings` index.
+
+The call returns a task ID to monitor the progress:
+
+[source,console]
+------------------------------------------------------------
+GET _tasks/<task_id> <1>
+------------------------------------------------------------
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc:287]:
```suggestion
GET _tasks/<task_id>
------------------------------------------------------------
//TEST[skip:TBD]
```

File: docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc [hunk 2]
```diff
@@ -172,10 +252,56 @@ POST _tasks/<task_id>/_cancel
...
+
+Reindexing large datasets can take a long time. You can test this workflow using only a subset of the dataset.
+
+To cancel the reindexing process and generate embeddings for the subset that was reindexed:
+
+[source,console]
+------------------------------------------------------------
+POST _tasks/<task_id>/_cancel <2>
+------------------------------------------------------------
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc:296]:
```suggestion
POST _tasks/<task_id>/_cancel
------------------------------------------------------------
//TEST[skip:TBD]
```

File: docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc [hunk 3]
```diff
@@ -282,6 +408,49 @@ query from the `semantic-embedding` index:
...
+                "query": "How to avoid muscle soreness while running?" 
+              }
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+------------------------------------------------------------
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-search-semantic-text.asciidoc:450]:
```suggestion
------------------------------------------------------------
// TEST[skip:TBD]
```

**Review by szabosteve — COMMENTED** [review]
These suggestions will make the CI pass. It's not a comprehensive review.

**Comment by leemthompo** [comment]
🚗 high-level impressions from me to help clarify this workflow for myself.

I think there should be just one index here, with a combination of `semantic_text` and standard `text` mappings. This eliminates a lot of redundancies, and simplifies the workflow. In my head this is how it should work, assuming we can also eliminate the reindexing step?

- Create an inference endpoint to set up the model for generating embeddings (e.g., ELSER)
- Set up **one** index with both standard `text` fields and `semantic_text` fields
- Index your documents, which will automatically generate embeddings for semantic_text fields
- Perform a semantic search query (existing content from tutorial)
- Perform hybrid search query by combining a standard `match` query (for example) and a `semantic` query, using RRF to combine the results from both query types

**Comment by szabosteve** [comment]
> Just a quick question about the reindexing step - if we use the API to index documents, I assume the embeddings will be generated automatically since the destination index mapping has the semantic_text field. Would that make reindexing unnecessary? Did I get that right?
However—please correct me if I'm wrong—this approach might only work for new documents. For already existing indices that users want to search on, wouldn't reindexing still be necessary to generate the embeddings? What do you think?

You are right. If you ingest documents – for example, by using the bulk API – then yes, the embeddings will be generated during ingest time. It makes the reindexing unnecessary. If you already have data in an index for which you want to create embeddings, you need to reindex that data into an index that has the semantic_text field, that's correct.

**Comment by szabosteve** [comment]
I think we should separate the hybrid search workflow to a standalone page – maybe a sub-page of the semantic text tutorial. One of the main reasons why we didn't demonstrate hybrid search on this page originally was to keep it relatively short, easy to follow, and demonstrate a simple workflow to get the idea and the machinery behind `semantic_search`. The content @kosabogi provided is very valuable and we need this in the docs. However, the presentation of the content in its current form makes the page busy and hard to follow which would make us fail on our first objective about simplicity.

I would suggest creating a page that demonstrates the `semantic_text` workflow with hybrid search and:

*  put it to the same level where the original tutorial sits
* or, put it one level deeper in the structure and link to it from the original tutorial.

I think the first solution would be better.

What do you think, @leemthompo, @kosabogi?

**Comment by leemthompo** [comment]
> However, the presentation of the content in its current form makes the page busy and hard to follow

I agree. Ultimately, I think we'll want to have clearer demarcation between tutorials, with a few full-text search tutorial, a few semantic search tutorials, and a few hybrid search tutorial. But we're missing some parts of that story today. 

So for right now, I agree that creating a separate hybrid search tutorial page is the right approach. It could be a sibling to semantic search or its own subsection, I think longer term it should be a sibling but it doesn't have to be from day one.

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 4]
```diff
@@ -0,0 +1,197 @@
+[[semantic-text-hybrid-search]]
+=== Tutorial: hybrid search with `semantic_text`
+++++
+<titleabbrev>Hybrid search with `semantic_text`</titleabbrev>
+++++
+
+This tutorial demonstrates how to perform **hybrid search**, combining semantic search with traditional full-text search.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:7]: Unnecessary highlighting.

```suggestion
This tutorial demonstrates how to perform hybrid search, combining semantic search with traditional full-text search.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    "num_allocations": 1,
+    "num_threads": 1
+  }
+}
+------------------------------------------------------------
+// TEST[skip:TBD]
+<1> The task type is `sparse_embedding` in the path as the `elser` service will
+be used and ELSER creates sparse vectors. The `inference_id` is
+`my-elser-endpoint`.
+<2> The `elser` service is used in this example.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:34]: It would be nice to provide a slightly more complex request to demonstrate adaptive resources, too.

```suggestion
[source,console]
------------------------------------------------------------
PUT _inference/sparse_embedding/my-elser-endpoint <1>
{
  "service": "elser", <2>
  "service_settings": {
    "adaptive_allocations": { <3>
      "enabled": true,
      "min_number_of_allocations": 3,
      "max_number_of_allocations": 10
    },
    "num_threads": 1
  }
}
------------------------------------------------------------
// TEST[skip:TBD]
<1> The task type is `sparse_embedding` in the path as the `elser` service will
be used and ELSER creates sparse vectors. The `inference_id` is
`my-elser-endpoint`.
<2> The `elser` service is used in this example.
<3> This setting enables and configures adaptive allocations.
Adaptive allocations make it possible for ELSER to automatically scale up or down resources based on the current load on the process.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+In hybrid search, semantic search retrieves results based on the meaning of the text, while full-text search focuses on exact word matches. By combining both methods, hybrid search delivers more relevant results, particularly in cases where relying on a single approach may not be sufficient.
+
+The recommended way to use hybrid search in the {stack} is following the `semantic_text` workflow. This tutorial uses the <<inference-example-elser,`elser` service>> for demonstration, but you can use any service and its supported models offered by the {infer-cap} API.
+
+[discrete]
+[[semantic-text-hybrid-infer-endpoint]]
+==== Create the {infer} endpoint
+
+Create an inference endpoint by using the <<put-inference-api>>:
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:17]: From 8.16, the default service for `semantic_text` is ELSER. You don't need to create an inference endpoint first, you just create an index with a semantic text field without specifying an inference_id and an inference endpoint with the ELSER service will be automagically created. This workflow will be demonstrated in the original semantic_text tutorial.

It is valuable to have the inference endpoint creation step documented here even if the example uses the default endpoint, because users can see what should be done in case they use a different service. However, we should mention that the default service is ELSER and we include this step only for demonstrating how the endpoint creation works.

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+This error usually just reflects a timeout, while the model downloads in the background.
+You can check the download progress in the {ml-app} UI.
+====
+
+[discrete]
+[[hybrid-search-create-index-mapping]]
+==== Create an index mapping for hybrid search
+
+The destination index (where the search will be performed) should contain both the embeddings for semantic search and the original text field for full-text search. This structure enables the combination of semantic search and full-text search, allowing the search engine to consider both the 
+meaning and the exact words in the query.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:48]:
```suggestion
The destination index will contain both the embeddings for semantic search and the original text field for full-text search. This structure enables the combination of semantic search and full-text search.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        "type": "text",
+        "copy_to": "semantic_text" <4>
+      }
+    }
+  }
+}
+------------------------------------------------------------
+// TEST[skip:TBD]
+<1> The name of the field to contain the generated embeddings for semantic search.
+<2> The `inference_id` is the inference endpoint that generates the embeddings based on the input text.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:70]:
```suggestion
<2> The identifier of the inference endpoint that generates the embeddings based on the input text.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+<3> The name of the field to contain the original text for lexical search.
+<4> The `copy_to` field copies the contents of the `content` field into the `semantic_text` field, enabling hybrid search functionality.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:72]: It is important to emphasize that in this case, the embeddings will be generated when the content is copied over from the `content` field to the `semantic_text` field by `copy_to`.

```suggestion
<4> The textual data stored in the `content` field will be copied to `semantic_text` and processed by the {infer} endpoint.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+Use the `msmarco-passagetest2019-top1000` data set, which is a subset of the MS MARCO Passage Ranking data set. It consists of 200 queries, each accompanied by a list of relevant text passages. All unique passages, along with their IDs, have been extracted from that data set and compiled into a https://github.com/elastic/stack-docs/blob/main/docs/en/stack/ml/nlp/data/msmarco-passagetest2019-unique.tsv[tsv file].
+
+Download the file and upload it to your cluster using the {kibana-ref}/connect-to-elasticsearch.html#upload-data-kibana[Data Visualizer] in the {ml-app} UI. After your data is analyzed, click **Override settings**. Under **Edit field names**, assign `id` to the first column and `content` to the second. Click **Apply**, then **Import**. Name the index `test-data`, and click **Import**. After the upload is complete, you will see an index named `test-data` with 182,469 documents.
+
+[discrete]
+[[hybrid-search-reindex-data]]
+==== Reindex the data for hybrid search
+
+Reindex the data from the `test-data` index into the `semantic-hybrid-embeddings` index to enable both lexical and semantic search.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:96]: I would avoid using `hybrid` in this case as it could suggest that the embeddings are different in case they are used in hybrid search while this is not the case.

```suggestion
Reindex the data from the `test-data` index into the `semantic-embeddings` index.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+The data in the `content` field of the source index is copied into the `content` field of the destination index. 
+The `copy_to` functionality ensures that the content is duplicated into the `semantic_text` field, where it will be processed by the inference endpoint to generate embeddings.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:98]:
```suggestion
The `copy_to` parameter set in the index mapping creation ensures that the content is copied into the `semantic_text` field. The data is processed by the {infer} endpoint at ingest time to generate embeddings.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+[source,console]
+------------------------------------------------------------
+POST _reindex?wait_for_completion=false
+{
+  "source": {
+    "index": "test-data",
+    "size": 10 <1>
+  },
+  "dest": {
+    "index": "semantic-hybrid-embeddings"
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:116]:
```suggestion
    "index": "semantic-embeddings"
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+------------------------------------------------------------
+POST _tasks/<task_id>/_cancel
+------------------------------------------------------------
+// TEST[skip:TBD]
+
+[discrete]
+[[hybrid-search-perform-search]]
+==== Perform hybrid search
+
+After reindexing the data into the `semantic-hybrid-embeddings` index, you can perform hybrid search, which combines both semantic and lexical search. You can perform hybrid search using <<rrf,reciprocal rank fusion (RRF)>>. RRF is a technique that merges the rankings from both semantic and lexical queries, giving more weight to results that rank high in either search. This ensures that the final results are balanced and relevant.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:147]:
```suggestion
After reindexing the data into the `semantic-embeddings` index, you can perform hybrid search by using <<rrf,reciprocal rank fusion (RRF)>>. RRF is a technique that merges the rankings from both semantic and lexical queries, giving more weight to results that rank high in either search. This ensures that the final results are balanced and relevant.
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+[source,console]
+------------------------------------------------------------
+GET semantic-hybrid-embeddings/_search
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:151]:
```suggestion
GET semantic-embeddings/_search
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+------------------------------------------------------------
+// TEST[skip:TBD]
+<1> The first `standard` retriever represents the traditional lexical search.
+<2> Lexical search is performed on the `content` field using the specified phrase.
+<3> The second `standard` retriever refers to the semantic search.
+<4> The `semantic_text` field is used to perform the semantic search.
+
+
+After performing the hybrid search, the query will return the top 10 documents that match both semantic and lexical search criteria. The results include detailed information about each document:
+
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:188]:
```suggestion
Would be helpful to include the relevant parts of the API response, similar to what we've done [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/semantic-search-semantic-text.html#semantic-text-semantic-search).
```

**Review by szabosteve — COMMENTED** [review]
It already looks great! 👏 
Left a couple of suggestions below.

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+[source,console]
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:194]:
```suggestion
[source,console-result]
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+    "successful": 1,
+    "skipped": 0,
+    "failed": 0
+  },
+  "hits": {
+    "total": {
+      "value": 473,
+      "relation": "eq"
+    },
+    "max_score": null,
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:210]: I think you can remove everything from 196 to 210 (both inclusive), so we can save some real estate

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+          },
+          "id": 8408852,
+          "content": "What so many out there do not realize is the importance of (...)" <5>
+        }
+      }
+    ]
+  }
+}
+------------------------------------------------------------
+// TEST[skip:TBD]
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:254]:
```suggestion
// NOTCONSOLE
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+  },
+  "hits": {
+    "total": {
+      "value": 473,
+      "relation": "eq"
+    },
+    "max_score": null,
+    "hits": [
+      {
+        "_index": "semantic-embeddings", <1>
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:213]:
```suggestion
        "_index": "semantic-embeddings",
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+        "_id": "wv65epIBEMBRnhfTsOFM", <2>
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:214]:
```suggestion
        "_id": "wv65epIBEMBRnhfTsOFM", 
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+        "_score": 0.032786883, <3>
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:215]:
```suggestion
        "_score": 0.032786883,
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+                    "##io": 1.0794281,
+                    "last": 1.0474665,
+                   (...) 
+                  }
+                }
+              ]
+            }
+          },
+          "id": 8408852,
+          "content": "What so many out there do not realize is the importance of (...)" <5>
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:247]:
```suggestion
          "content": "What so many out there do not realize is the importance of (...)"
```

File: docs/reference/search/search-your-data/semantic-text-hybrid-search [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+    ]
+  }
+}
+------------------------------------------------------------
+// TEST[skip:TBD]
+<1> The name of the index the document belongs to.
+<2> The unique identifier of the document.
+<3> The relevance score of the document based on the search query.
+<4> The processed embeddings generated by the inference endpoint.
+<5> The original text, allowing for full-text search.
```
Comment (szabosteve) [inline docs/reference/search/search-your-data/semantic-text-hybrid-search:259]:
```suggestion
```

**Comment by szabosteve** [comment]
@elasticsearchmachine test this please

**Review by szabosteve — APPROVED** [review]
LGTM!
